server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalGeoHashGrid.java

```diff
@@ -63,26 +65,29 @@
          * Read from a stream.
          */
         private Bucket(StreamInput in) throws IOException {
+            type = GeoHashType.readFromStream(in);
             geohashAsLong = in.readLong();
             docCount = in.readVLong();
             aggregations = InternalAggregations.readAggregations(in);
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            type.writeTo(out);
             out.writeLong(geohashAsLong);
             out.writeVLong(docCount);
             aggregations.writeTo(out);
         }
 
         @Override
         public String getKeyAsString() {
-            return GeoHashUtils.stringEncode(geohashAsLong);
+            return type.getHandler().hashAsString(geohashAsLong);
         }
 
         @Override
         public GeoPoint getKey() {
-            return GeoPoint.fromGeohash(geohashAsLong);
+            // TODO/FIXME:  is it ok to change from GeoPoint to Object, and return different types?
```

Comment:
Can you explain what you mean here?

Reply:
I am on the fence here but leaning towards keeping GeoPoint. The reason for this is that I think its good to have a consistent Class for the key so users can change the type and always know what to expect back from this method. Having different Classes returned from different types would mean that a client would need to consult the type in order to know what to do with the object returned here which would make their code much more complex.